gateway/handler_test.go

```diff
@@ -26,3 +37,75 @@ func TestEtagMatch(t *testing.T) {
 		}
 	}
 }
+
+type errorMockAPI struct {
+	err error
+}
+
+func (api *errorMockAPI) GetUnixFsNode(context.Context, ipath.Resolved) (files.Node, error) {
+	return nil, api.err
+}
+
+func (api *errorMockAPI) LsUnixFsDir(ctx context.Context, p ipath.Resolved) (<-chan iface.DirEntry, error) {
+	return nil, api.err
+}
+
+func (api *errorMockAPI) GetBlock(ctx context.Context, c cid.Cid) (blocks.Block, error) {
+	return nil, api.err
+}
+
+func (api *errorMockAPI) GetIPNSRecord(ctx context.Context, c cid.Cid) ([]byte, error) {
+	return nil, api.err
+}
+
+func (api *errorMockAPI) GetDNSLinkRecord(ctx context.Context, hostname string) (ipath.Path, error) {
+	return nil, api.err
+}
+
+func (api *errorMockAPI) IsCached(ctx context.Context, p ipath.Path) bool {
+	return false
+}
+
+func (api *errorMockAPI) ResolvePath(ctx context.Context, ip ipath.Path) (ipath.Resolved, error) {
+	return nil, api.err
+}
+
+func TestGatewayTimeoutBubblingFromAPI(t *testing.T) {
+	api := &errorMockAPI{err: fmt.Errorf("the mock api has timed out: %w", ErrGatewayTimeout)}
+	ts := newTestServer(t, api)
+	t.Logf("test server url: %s", ts.URL)
+
+	req, err := http.NewRequest(http.MethodGet, ts.URL+"/ipns/ipfs.io", nil)
```

Comment:
nit: remember to use different name than `ipfs.io`:  using `ipfs` in the past  hid some errors because of code thinking it is  `/ipfs/` namespace, better safe than sorry and use something else (`example.net`, or `en.wikipedia-on-ipfs.org`)


```suggestion
	req, err := http.NewRequest(http.MethodGet, ts.URL+"/ipns/en.wikipedia-on-ipfs.org", nil)
```